BootloaderCorePkg/Library/PciEnumerationLib/PciEnumerationLib.c

```diff
@@ -1136,43 +1144,62 @@ PciEnumeration (
   IN  VOID   *MemPool
   )
 {
+  EFI_STATUS                 Status;
   UINT32                     Idx;
-  UINT8                      SubBusNumber;
+  UINT16                     PrimaryBus;
+  UINT16                     SubBusNumber;
+  UINT64                     IoBase;
+  UINT64                     Mem32Base;
   PCI_IO_DEVICE             *Root;
 
   SetAllocationPool (MemPool);
 
-  SubBusNumber = 0;
-  Root = CreatePciIoDevice (NULL, NULL, 0, 0, 0);
-  PciScanBus (Root, 0, &SubBusNumber, NULL);
+  IoBase = PcdGet32 (PcdPciResourceIoBase);
+  Mem32Base = PcdGet32 (PcdPciResourceMem32Base);
+  for (PrimaryBus = 0; PrimaryBus <= PCI_MAX_BUS; PrimaryBus = SubBusNumber + 1) {
+    SubBusNumber = PrimaryBus;

```

Comment:
Not sure the boot performance impact to core platform. If big impact, maybe we could add a policy to decide if we need  scan all the possible buses.

Reply:
We can use Pcd to limit max bus iteration. How do you think?
One concern is we need to update start/end bus number in MCFG table. But, all boards has hard-coded value 0 - 0xFF currently. For MCFG table, whole iteration is required and the MCFG field needs to be updated with the detected bridges. This is also TBD.

Reply:
Did you get the boot perf impact for this patch? If the impact is obvious, yes, we could add a PCD for max bus number.

Reply:
Let me check perf number.

Reply:
There is huge gap in perf number - 3ms -> 85ms. Let me define a Pcd to limit the number of bus iteration.

Reply:
OK. make the default value to same with the current behavior for core platform.

Reply:
I think the better solution is to align wiht EDK2 to use root bridge concept.   In this way, we don't have to scan the non-existing buses.  Each root bridge will have a start bus.  The PCI scan bus will start from this bus.

Reply:
Let me try to scan only root bridge first, then iterate the bridges and compare perf number.